content/recipes/async-local-storage.md

```diff
@@ -0,0 +1,227 @@
+### Async Local Storage
+
+`AsyncLocalStorage` is a [Node.js API](https://nodejs.org/api/async_context.html#async_context_class_asynclocalstorage) (based on the `async_hooks` API) that provides an alternative way of propagating local state through the application without the need to explicitly pass it as a function parameter. It is similar to a thread-local storage in other languages.
+
+The main idea of Async Local Storage is that we can _wrap_ some function call with the `AsyncLocalStorage#run` call. All code that is invoked within the wrapped call gets access to the same `store`, which will be unique to each call chain.
+
+In the context of NestJS, that means if we can find a place within the request's lifecycle where we can wrap the rest of the request's code, we will be able to access and modify state visible only to that request, effectively eliminating the need for REQUEST scoped providers and their limitations.
+
+#### Custom implementation
+
+NestJS itself does not provide any built-in abstraction for `AsyncLocalStorage`, so let's walk through how we could implement it ourselves for the simplest HTTP case to get a better understanding of the whole concept:
+
+> info **info** For a ready-made [dedicated package](recipes/async-local-storage#nestjs-cls), continue reading below.
+
+1. First, create a new instance of the `AsyncLocalStorage` in some shared source file. Since we're using NestJS, let's also turn it into a module with a custom provider.
+
+```ts
+/** als.module.ts */
+
```

Comment:
```suggestion
@@filename(als.module)
```

Reply:
This causes the build to fail with this error
![image](https://user-images.githubusercontent.com/46406259/197463037-3d5e2914-6a32-4594-81f5-4dc94b8fe943.png)
Do you know what might be the cause?

Reply:
It's also strange, because the same syntax is used in `cqrs.component.ts`, which is part of the same `recipes.module.ts` as `async-local-storage.component.ts`

Reply:
What if we change `ts` to `typescript`

Reply:
No change. I even tried copy-and-pasting an entire block from the `cqrs.ts` file and the error is still there. :disappointed:

Reply:
I even tried putting the `ExtensionPipe` directly in the `RecipesModule`, but that only caused more errors. That's where my angular debugging proficiency ends. @jmcdo29, would you have any idea what might be wrong?

Reply:
Huh, I'll have to work with this locally and try to figure out what that's about

Reply:
You need to add `ASyncLocalStorageComponents` to the `RecipeModule`'s `declarations`

```diff                                                   
      ServeStaticComponent,                                                              
      NestCommanderComponent,                                                            
      ReplComponent,                                                                     
+     AsyncLocalStorageComponent,                                                        
    ],                                                                                   
  })                                                                                     
- export class RecipesModule {}                                                          
+ export class RecipesModule { }                                                                                                                                                
```